Create a section for the 0.80 philosophy (#656)
nushell/nushell.github.io: contributor-book/philosophy_0_80.md

# Nushell design philosophy

## Core philosophy

Nushell is "A shell-first scripting language for working with structured data flowing through pipelines".

Designs should work towards serving this goal. Those that don't should be removed or moved to optional add-ons.

## Core design

### Shell-first

Nushell should work to serve its roll as a shell and a language with a focus on shells. Shell types of activities which include, but aren't limited to:

* Running commands
* Redirecting stdout/stdin/stderr
* Unix-only: Properly handling signals
    * Ctrl-C
    * Ctrl-D
    * and others
* Unix-only: Handling background tasks
    * Ctrl-Z

>     * Background/foregrounding of tasks

Reviewer: I'm generally agree with @dandavison

Additionally, it seems the only way to manage task is spawning task by multi-processing, or we have no way to cancel task(we can't cancel a thread).  And it requires our whole engine-state to be serializable/deserializable.  Currently I found `pueue` and `terminal multiplexers` are working really fine to me.   For Ctrl-Z, sometimes I'm using this, but I don't think it's necessary either.

>     * Background/foregrounding of tasks

Reviewer 2: This is probably the thing I miss most from bash, but I've found work arounds. Ctrl-z is great for backgrounding neovim and resuming without having to restart the LSP server, which for rust-analyzer has an annoyingly long startup time.

I use kitty these days for a quick new shell session, and there are neovim terminal plugins but the downside is they all take you out to a different window or force you into a smaller screen area

>     * Background/foregrounding of tasks

Reviewer 2: > I'm generally agree with @dandavison
> 
> Additionally, it seems the only way to manage task is spawning task by multi-processing, or we have no way to cancel task(we can't cancel a thread). And it requires our whole engine-state to be serializable/deserializable. Currently I found `pueue` and `terminal multiplexers` are working really fine to me. For Ctrl-Z, sometimes I'm using this, but I don't think it's necessary either.

My only gripe with pueue so far is interactivity. Having to send commands through pueue instead of just reattaching a shell to the process is unergonomic

>     * Background/foregrounding of tasks

Author: I'll turn this into a more general "support for background tasks" as we'll probably want some solution that works reasonably well for them and hopefully one that could work crossplatform.

In cases where the user would reasonably have an expectation of the functionality of the shell, we should make every effort to include. If a feature would be a reasonable expectation (say ctrl-z on Unix), then we should have OS-specific functionality for that platform that meets that expectation.

### Scripting language

The intent of Nushell's design is to be a scripting language. Scripting can take many forms, from simple pipelines to large scripts. Nushell should handle these with ease by meeting reasonable expectations in terms of:
- Modularity
- Readability
- Programming language features and convenience\

### Structured data

> In Nushell, all data is "structured". For Nushell, this means that values can take shapes beyond just a simple string.

Reviewer 3: I think of the type information as providing structure.

> In Nushell, all data is "structured". For Nushell, this means that values can take shapes beyond just a simple string.

Reviewer 2: Agreed. In other shells, everything is just “string". Nushell at least provides a distinction while still flexible enough to fall back to String or raw binary when the structured type isn't known

- The name of the command
- The usage information (documentation) for the command
- The name and type of:
    - Named parameters
    - Positional parameters
- The type of:
    - Input
    - Output

### When to use input

Input is intended for one or both of:
- Pipeline composition
- Streams of data

### When to use output

All return values are output.

### When to use positional parameters

Use a positional parameter when:
- A parameter is required

### When to use rest parameters

Use a rest parameter when:
- A command takes some number of additional optional arguments of the same type

### When to use named parameters

Use a named parameter (a flag parameter) when:
- The value you need is optional

### When to use a switch flag

Use a switch flag when:
- You need to allow the user to change the default action of the command

### Comments

Commands should be commented to describe their function and usage. This documentation should also cover parameters and their use.

## Core categories

The core language and standard library needs to cover the following categories to support common use cases for Nushell:

* Filesytem
* Operating system
* Manipulating the environment

> * Parsing string data into structure data

Reviewer 4: ```suggestion
* Parsing string data into structure data
* Formatting structured data as standard string formats (CSV and JSON)
```

Suggestion is to replace the CSV/JSON entry below with this.

* Querying, filtering, and manipulating structured data
* Network connectivity
    * note: Network support is fundamental because with it users can easily acquire and install Nushell extensions
* Basic formats:

>     * CSV and JSON support

Reviewer 3: I think a case could be made for keeping TOML/YAML/XML in the core as well. But I'd be open to just CSV+JSON.

Reviewer 5: +1 for TOML

Reviewer: I think we can have more format supported, in additional to @rgwood mentioned, I'd like to have `ssv`, `ini`, 'xlsx` supports.   But these would introduce too much dependencies into nushell core itself, I'm not sure...

Author: I changed it so the exact formats are "to be decided" rather than listing them now. We should look to see which ones are used a lot in practice in current Nushell.

* Basic date support

> The following categories should be moved to plugins:

Reviewer 4: Yes, I think so! I know this is a big change, but I think it has the potential to move us into a new era of a v2 builtins stdlib that is really starting to be on-course for 1.0.

I'm not super familiar with the plugin technology. As an aside, eventually it will be appropriate for people to implement libraries of utilities for the public in nushell itself rather than Rust plugins.

* Database connectivity
* Dataframe support
* Hash functionality
* Format support
    * Any non-CSV or JSON file format
* Experimental commands
* Binary data functionality
* Random number support

> * Math support

Reviewer 3: IMO these 3 feel like they should be in core. If Nu can't do math or work with binary data out of the box, that feels like a shortcoming.

* Advanced date support

We should select commands to be in the core categories which meet the common use cases for Nushell. Commands that are in the core categories that are uncommon use cases should move to optional extensions.

## Language design

Rather than describe the whole of the language here, this section describes the changes expected to come as part of 0.80 and how they differ from the 0.60 series.

### More bash-ism

Being a shell-focused language means incorporating more of the expected shell-features into the language. These include:

Redirection (with common variations):
```
cat foo.txt > bar.txt
cat foo.txt >> bar.txt
cat foo.txt 2> bar.txt
```

Bash logic operators:
```
cat foo.txt && cat bar.txt
cat foo.txt || cat bar.txt

> ```

Reviewer 4: I definitely agree with those two categories of bashisms entering the language.

I wouldn't want to see much more though without plenty of justifying design discussion.

Author: Agreed - I think we want to catch the most common bash-isms without trying for some kind of full bash-compatibility.

Reviewer 2: I'm with @dandavison partially. The redirection operators from bash are notoriously difficult to use correctly, but some thought on how to do it in a nutshell way would be useful.

It's very difficult to redirect certain things in nushell that bash can do without resorting to just jumping down into bash with a do block. Specifically things like infinite streams of stderr, or splitting stderr and stdout to different files at the same time, teeing one of them, etc

Author: Added some wording to limit this to just the bashisms described.

* Running commands
* Redirecting stdout/stdin/stderr
* Unix-only: Properly handling signals
    * Ctrl-C
    * Ctrl-D
    * and others
* Unix-only: Handling background tasks
    * Ctrl-Z
    * Background/foregrounding of tasks

In cases where the user would reasonably have an expectation of the functionality of the shell, we should make every effort to include. If a feature would be a reasonable expectation (say ctrl-z on Unix), then we should have OS-specific functionality for that platform that meets that expectation.

### Scripting language

The intent of Nushell's design is to be a scripting language. Scripting can take many forms, from simple pipelines to large scripts. Nushell should handle these with ease by meeting reasonable expectations in terms of:
- Modularity
- Readability
- Programming language features and convenience

### Structured data

In Nushell, all data is "structured". For Nushell, this means that values can take shapes beyond just a simple string.

Values in Nushell can be records, lists, tables, binary data, and more. Being able to convert into structured data and work with structured data is fundamental to Nushell.

### Pipelines

> Nushell takes the Unix philosophy of pipelines to heart. Commands should be built with the intent of composition. Nushell leverages this composition via the use of pipes (`|`), just like Unix pipelines.

Reviewer 4: ```suggestion
Nushell takes the Unix philosophy of pipelines to heart. Commands should be built with the intent of composition. Nushell enables composition via the use of pipes (`|`), just like Unix pipelines.
```

Composing commands, both built-in and user-defined, is a core piece of Nushell. The design of Nushell and its standard library must support both easily composing commands as well as allowing the user to easily create compose-able commands.

> 

Reviewer 4: ```suggestion
### Command philosophy

Specifically, Nushell's philosophy about commands is described by the following positions:

1. There should be one -- and preferably only one -- obvious way to do it.
  You will probably recognize this from the "[Zen of Python](https://peps.python.org/pep-0020/)"!  

2. If something can be done conveniently by composition of simpler commands then we do not add flags and options to do the same thing.

3. The primary input to a command should generally be supplied as "input", not as a positional argument.
  This is what enables pipeline composition.

4. A given command, on a given input, should generally always produce the same type of output data structure; the presence/absence/value of arguments and flags shouldn't change the output type.
```

Reviewer 4: Sorry I dumped these here without commenting. They're just a quickly written-down attempt at some command philosophy.

Author: I'll land them as is. I wrote the other text pretty loosely as well and I'm assuming we'll continue refining it before we land it.

# Nushell design philosophy

## Core philosophy

Nushell is "A shell-first scripting language for working with structured data flowing through pipelines".

Designs should work towards serving this goal. Those that don't should be removed or moved to optional add-ons.

## Core design

### Shell-first

> Nushell should work to serve its roll as a shell and a language with a focus on shells. Shell types of activities include, but aren't limited to:

Reviewer 6: ```suggestion
Nushell should work to serve its role as a shell and a language with a focus on shells. Shell types of activities include, but aren't limited to:
```

## Command signatures and their parts

### Signature

The signature of the command describes the following:

- The name of the command
- The usage information (documentation) for the command
- The name and type of:
    - Named parameters
    - Positional parameters
- The type of:
    - Input
    - Output

### When to use input

Input is intended for one or both of:
- Pipeline composition
- Streams of data

### When to use output

All return values are output.

### When to use positional parameters

Use a positional parameter when:
- A parameter is required

### When to use rest parameters

Use a rest parameter when:
- A command takes some number of additional optional arguments of the same type

### When to use named parameters

Use a named parameter (a flag parameter) when:
- The value you need is optional

### When to use a switch flag

Use a switch flag when:
- You need to allow the user to change the default action of the command

### Comments

Commands should be commented to describe their function and usage. This documentation should also cover parameters and their use.

> 

Reviewer 4: ```suggestion
## Naming commands

Good naming is incredibly important! Here are some guidelines:

- Find out what names other well-known programming languages use for the thing you are creating (if something similar exists). Don't choose a name that differs from all other languages without a good reason.

- As always in programming, if you are uncomfortable with the name chosen for something, stop! It often means there's something wrong with your design. Maybe it's doing too much? Or maybe the concept it's representing is still a bit confused?

- Don't expect it to be easy -- it may take some time to settle on the right name for a command (and its options, if it needs any). Don't hesitate to come and talk to other Nushell contributors in Discord, or discuss it in GitHub.

- Find out what names Nushell uses for related things -- your name should be harmonious with those. E.g. if `head` already exists, maybe it makes sense for another command called `tail` to exist? (Or maybe the thing called `head` shouldn't have that name! Naming is hard so don't assume current names are perfect or immutable.)

- Remember that when you're choosing a name, you're making a decision that affects the code of every Nushell user in the world. (Of course, this means that fixing bad naming decisions will break the code of Nushell users and force them to update. We still won't always get it right the first time, but this is one reason that we put so much emphasis on it.)

- If the name you feel is the "right name" is already taken, don't just settle for an inferior name -- please start a discussion about the situation instead. It may mean that there is a broader conceptual or naming problem that needs to be addressed.

- Please explain how you chose the name in your PR -- writing this will help with the process of naming it!

The ultimate goal here is to create an ecosystem of commands and terminology that is harmonious, and "feels right". Yes, this is both highly subjective and hard work! But it is an extremely important factor determining how enjoyable and powerful it is to work in the programming environment that we are creating, and ultimately it is this that will determine the success of the project.
```

Reviewer 7: I agree with the sentiment here. Indeed, naming can be difficult, but REnaming is frequently trivial. I'd rather see some awesome wow thing that changes the way we think about things, that is named badly, then have folks struggling for hours trying to come up with the "right" name and never really starting to code. Try to do all the things above but don't stop coding because you don't have the perfect name for something.

Here's a real-world example. zhiburt, who is the author of tabled and a frequent contributor to nushell recently introduced a PR to show off some cool tui things with horizontal and vertical scrolling. The initial name was tabless, kind of clever like table and less combined, I get it. But the point of the PR wasn't the name, it was the total awesomeness that can come with a tui. That thing could've been named bob initially and I would've been equally blown away with it. It truly was a beautiful demo and surprise. I tried to encourage him, and we go back and forth about features and he kept on chugging away. Then someone objected to the name and suggested scroll. zhiburt changed it in a few minutes and scroll is what it is currently named now. Is that the perfect name? I have no idea, but tbh, I'm not really concerned with the exact "right" name. I want to use the tui he introduced whatever its final name ends up being. Having said that, it's sometimes easier to spot the wrong name than to pick the right one. bob would've been the wrong name. lol. And yet again, sometimes other people can just come up with good names, but you can't because you can't see the forest for the trees.

Reviewer 7: I agree with the sentiment here. Indeed, naming can be difficult, but REnaming is frequently trivial. I'd rather see some awesome wow thing that changes the way we think about things, that is named badly, then have folks struggling for hours trying to come up with the "right" name and never really starting to code. Try to do all the things above but don't stop coding because you don't have the perfect name for something.

Here's a real-world example. zhiburt, who is the author of tabled and a frequent contributor to nushell recently introduced a PR to show off some cool tui things with horizontal and vertical scrolling. The initial name was tabless, kind of clever like table and less combined, I get it. But the point of the PR wasn't the name, it was the total awesomeness that can come with a tui. That thing could've been named bob initially and I would've been equally blown away with it. It truly was a beautiful demo and surprise. I tried to encourage him, and we go back and forth about features and he kept on chugging away. Then someone objected to the name and suggested scroll. zhiburt changed it in a few minutes and scroll is what it is currently named now. Is that the perfect name? I have no idea, but tbh, I'm not really concerned with the exact "right" name. I want to use the tui he introduced whatever its final name ends up being. Having said that, it's sometimes easier to spot the wrong name than to pick the right one. bob would've been the wrong name. lol. And yet again, sometimes other people can just come up with good names, but you can't because you can't see the forest for the trees.

Reviewer 4: Thanks @fdncred, I agree we should iterate on these guidelines together. It is true that renaming is trivial when something's not merged. However it's still true that if you are struggling with naming something then that often points to a problem with the design of what you're working on.

But also, what I was getting at (and which I didn't make clear) is that renaming is not at all trivial when something is merged and released. So the purpose of my bullet points was to emphasize that PRs should not be merged until multiple people are happy with the naming. Currently, there are many bad naming and API decisions in the library of Nushell builtins, but those builtins are out there in the wild being used by Nushell users and perhaps even playing a role in working software. So we'll need to make breaking changes to fix this, but we also need to change our own processes so that we get better at releasing builtins that don't need breaking changes to fix design decisions.

Author: @dandavison - we'll definitely need to set up naming guidelines so we can get commands names in a predictable/readable shape for the 1.0 release. I think it might take a few revisions to get it how we want it (since I'm sure we even know just yet 😅 )

Can we split the naming piece into its own philosophy doc? That way we can land this one sooner and then add more components to it one after another as they get ready.

Reviewer 4: > Can we split the naming piece into its own philosophy doc?

Absolutely! I just wanted to get some thoughts and draft text down on the subject for others to comment on.

## Core categories

The core language and standard library needs to cover the following categories to support common use cases for Nushell:

* Filesytem
* Operating system
* Manipulating the environment
* Parsing string data into structured data
* Formatting structured data as standard string formats (CSV and JSON)
* Querying, filtering, and manipulating structured data
* Network connectivity
    * note: Network support is fundamental because with it users can easily acquire and install Nushell extensions
* Basic formats:
    * CSV and JSON support
* Basic date support

The following categories should be moved to plugins:

* Database connectivity
* Dataframe support

> * Hash functionality

Reviewer: I think this should keep in core too.  In bash I usually using md5sum to calculate a downloaded file to verify it's not corrupted.
